File: mobile-ios/Modules/Domain/Sources/Entities/Order.swift
import Foundation

public enum OrderStatus: String, Sendable, CaseIterable {
    case pending
    case confirmed
    case preparing
    case outForDelivery = "out_for_delivery"
    case delivered
    case cancelled

    public var displayName: String {
        switch self {
        case .pending: "Sipariş Alındı"
        case .confirmed: "Onaylandı"
        case .preparing: "Hazırlanıyor"
        case .outForDelivery: "Yolda"
        case .delivered: "Teslim Edildi"
        case .cancelled: "İptal Edildi"
        }
    }

    /// Müşteri yalnızca hazırlığa başlanmadan iptal edebilir.
    public var isCancellableByCustomer: Bool {
        self == .pending || self == .confirmed
    }

    /// Hâlâ süren sipariş (teslim/iptal değil) — ana ekran takip kartı bunları gösterir.
    public var isActive: Bool {
        self != .delivered && self != .cancelled
    }

    /// Zaman çizelgesinde gösterilen normal akış.
    public static let deliveryFlow: [OrderStatus] =
        [.pending, .confirmed, .preparing, .outForDelivery, .delivered]
}

public enum PaymentMethod: String, Sendable, CaseIterable {
    case cashOnDelivery = "cash_on_delivery"
    case card

    public var displayName: String {
        switch self {
        case .cashOnDelivery: "Kapıda nakit"
        case .card: "Kapıda kart"
        }
    }
}

public struct OrderLine: Identifiable, Equatable, Sendable {
    public let id: String
    public let productID: String
    public let name: String
    public let quantity: Int
    public let unitPrice: Decimal
    public let lineSubtotal: Decimal

    public init(
        id: String,
        productID: String,
        name: String,
        quantity: Int,
        unitPrice: Decimal,
        lineSubtotal: Decimal
    ) {
        self.id = id
        self.productID = productID
        self.name = name
        self.quantity = quantity
        self.unitPrice = unitPrice
        self.lineSubtotal = lineSubtotal
    }
}

public struct OrderEvent: Identifiable, Equatable, Sendable {
    public let id: String
    public let status: OrderStatus
    public let note: String?
    public let createdAt: Date?

    public init(id: String, status: OrderStatus, note: String?, createdAt: Date?) {
        self.id = id
        self.status = status
        self.note = note
        self.createdAt = createdAt
    }
}

public struct Order: Identifiable, Equatable, Sendable {
    public let id: String
    public let status: OrderStatus
    public let paymentMethod: PaymentMethod
    public let subtotal: Decimal
    public let discountAmount: Decimal
    public let totalAmount: Decimal
    public let createdAt: Date?
    public let lines: [OrderLine]
    public let statusHistory: [OrderEvent]
    public let address: Address?
    /// Yalnızca admin sipariş listesinde dolu (müşteriye açık uçlarda nil).
    public let customerName: String?
    public let customerPhone: String?

    public init(
        id: String,
        status: OrderStatus,
        paymentMethod: PaymentMethod,
        subtotal: Decimal,
        discountAmount: Decimal,
        totalAmount: Decimal,
        createdAt: Date?,
        lines: [OrderLine],
        statusHistory: [OrderEvent],
        address: Address?,
        customerName: String? = nil,
        customerPhone: String? = nil
    ) {
        self.id = id
        self.status = status
        self.paymentMethod = paymentMethod
        self.subtotal = subtotal
        self.discountAmount = discountAmount
        self.totalAmount = totalAmount
        self.createdAt = createdAt
        self.lines = lines
        self.statusHistory = statusHistory
        self.address = address
        self.customerName = customerName
        self.customerPhone = customerPhone
    }

    public var itemCount: Int {
        lines.reduce(0) { $0 + $1.quantity }
    }

    /// Kısa referans (id'nin ilk bloğu).
    public var reference: String {
        String(id.prefix(8)).uppercased()
    }

    /// Admin için bir sonraki durum (ileri akış). Teslim/iptal ise nil.
    public var nextStatus: OrderStatus? {
        let flow = OrderStatus.deliveryFlow
        guard let i = flow.firstIndex(of: status), i < flow.count - 1 else { return nil }
        return flow[i + 1]
    }
}

/// Sipariş oluşturma girdisi.
public struct PlaceOrderInput: Sendable, Equatable {
    public var addressID: String
    public var items: [(productID: String, quantity: Int)]
    public var couponCode: String?
    public var paymentMethod: PaymentMethod

    public init(
        addressID: String,
        items: [(productID: String, quantity: Int)],
        couponCode: String?,
        paymentMethod: PaymentMethod
    ) {
        self.addressID = addressID
        self.items = items
        self.couponCode = couponCode
        self.paymentMethod = paymentMethod
    }

    public static func == (lhs: PlaceOrderInput, rhs: PlaceOrderInput) -> Bool {
        lhs.addressID == rhs.addressID
            && lhs.couponCode == rhs.couponCode
            && lhs.paymentMethod == rhs.paymentMethod
            && lhs.items.count == rhs.items.count
            && zip(lhs.items, rhs.items)
            .allSatisfy { $0.productID == $1.productID && $0.quantity == $1.quantity }
    }
}

File: mobile-android/app/src/main/java/com/erenlermarket/app/domain/model/Order.kt
package com.erenlermarket.app.domain.model

import java.math.BigDecimal

enum class OrderStatus(val apiValue: String) {
    PENDING("pending"),
    CONFIRMED("confirmed"),
    PREPARING("preparing"),
    OUT_FOR_DELIVERY("out_for_delivery"),
    DELIVERED("delivered"),
    CANCELLED("cancelled"),
    ;

    val displayName: String
        get() = when (this) {
            PENDING -> "Sipariş alındı"
            CONFIRMED -> "Onaylandı"
            PREPARING -> "Hazırlanıyor"
            OUT_FOR_DELIVERY -> "Yolda"
            DELIVERED -> "Teslim edildi"
            CANCELLED -> "İptal edildi"
        }

    /** Müşteri yalnızca hazırlığa başlanmadan iptal edebilir. */
    val isCancellableByCustomer: Boolean get() = this == PENDING || this == CONFIRMED

    companion object {
        fun from(value: String?): OrderStatus =
            entries.firstOrNull { it.apiValue == value } ?: PENDING

        /** Zaman çizelgesinde gösterilen normal akış. */
        val deliveryFlow = listOf(PENDING, CONFIRMED, PREPARING, OUT_FOR_DELIVERY, DELIVERED)
    }
}

enum class PaymentMethod(val apiValue: String) {
    CASH_ON_DELIVERY("cash_on_delivery"),
    CARD("card"),
    ;

    val displayName: String
        get() = when (this) {
            CASH_ON_DELIVERY -> "Kapıda nakit"
            CARD -> "Kapıda kart"
        }

    companion object {
        fun from(value: String?): PaymentMethod =
            entries.firstOrNull { it.apiValue == value } ?: CASH_ON_DELIVERY
    }
}

data class OrderLine(
    val id: String,
    val productId: String,
    val name: String,
    val quantity: Int,
    val unitPrice: BigDecimal,
    val lineSubtotal: BigDecimal,
)

data class OrderEvent(
    val id: String,
    val status: OrderStatus,
    val note: String?,
    val createdAt: String?,
)

data class Order(
    val id: String,
    val status: OrderStatus,
    val paymentMethod: PaymentMethod,
    val subtotal: BigDecimal,
    val discountAmount: BigDecimal,
    val totalAmount: BigDecimal,
    val createdAt: String?,
    val lines: List<OrderLine>,
    val statusHistory: List<OrderEvent>,
    val address: Address?,
) {
    val itemCount: Int get() = lines.sumOf { it.quantity }

    /** Kısa referans (id'nin ilk bloğu). */
    val reference: String get() = id.take(8).uppercase()
}

/** Sipariş oluşturma girdisi. */
data class PlaceOrderInput(
    val addressId: String,
    val items: List<Item>,
    val couponCode: String?,
    val paymentMethod: PaymentMethod,
) {
    data class Item(val productId: String, val quantity: Int)
}
